### PR TITLE
tracing(aws): add configuration option to inject all messages with trace context during SQS, SNS, and Kinesis batch send operations

### PR DESCRIPTION
**Pull request by wconti27**
### What does this PR do?
Adds an `aws-sdk` plugin config option to inject all AWS messages during batch send operations with trace context (SQS, SNS, Kinesis).
The config option can be enabled for all three services using the following:
`tracer.use('aws-sdk', { kinesis: { batchPropagationEnabled: true }, sns: { batchPropagationEnabled: true }, sqs: { batchPropagationEnabled: true } })`

### Motivation
Customer request

### Plugin Checklist

- [x] Unit tests.
- [x] TypeScript [definitions][1].
- [x] TypeScript [tests][2].
- [x] API [documentation][3].
- [x] CircleCI [jobs/workflows][4].
- [x] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/documentation/blob/master/content/en/tracing/trace_collection/library_config/nodejs.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

**Review by tabgok — DISMISSED** [review]
Approving from a functionality standpoint

**Review by Qard — DISMISSED** [review]

File: packages/datadog-plugin-aws-sdk/src/services/kinesis.js [hunk 1]
```diff
@@ -152,7 +152,12 @@ class Kinesis extends BaseAwsSdkPlugin {
       case 'putRecords':
         stream = params.StreamArn ? params.StreamArn : (params.StreamName ? params.StreamName : '')
         for (let i = 0; i < params.Records.length; i++) {
-          this.injectToMessage(span, params.Records[i], stream, i === 0)
+          this.injectToMessage(
+            span,
+            params.Records[i],
+            stream,
+            i === 0 || (this.config.kinesis && this.config.kinesis.batchPropagationEnabled)
```
Comment (tlhunter) [inline packages/datadog-plugin-aws-sdk/src/services/kinesis.js:159]: Can you hoist this just before the for loop?

```js
const enableKinesisBatch = this.config.kinesis && this.config.kinesis.batchPropagationEnabled
// ...
i === 0 || enableKinesisBatch
```

File: packages/datadog-plugin-aws-sdk/src/services/sns.js [hunk 2]
```diff
@@ -59,7 +59,12 @@ class Sns extends BaseAwsSdkPlugin {
         break
       case 'publishBatch':
         for (let i = 0; i < params.PublishBatchRequestEntries.length; i++) {
-          this.injectToMessage(span, params.PublishBatchRequestEntries[i], params.TopicArn, i === 0)
+          this.injectToMessage(
+            span,
+            params.PublishBatchRequestEntries[i],
+            params.TopicArn,
+            i === 0 || (this.config.sns && this.config.sns.batchPropagationEnabled)
```
Comment (tlhunter) [inline packages/datadog-plugin-aws-sdk/src/services/sns.js:66]: Same hoisting here

**Comment by tlhunter** [comment]
What are the drawbacks for having this always enabled? If there are none, shouldn't we always have it enabled? If there is a drawback, shouldn't the API docs for the configuration call this out?

Are batches only used when user code manually calls a batch method? Or does the AWS library decide to batch some things?
I'm curious if two requests to a Node server come in at the same time, one with request ID 1, one with request ID 2, and both making an outbound AWS request. Does the AWS SDK internally every combine those into a batch? And if so, would the request IDs be properly associated?

File: packages/datadog-plugin-aws-sdk/src/services/kinesis.js (hunk 1 above)
Comment (wconti27) [reply packages/datadog-plugin-aws-sdk/src/services/kinesis.js:159]: done

**Comment by wconti27** [comment]
> What are the drawbacks for having this always enabled? If there are none, shouldn't we always have it enabled? If there is a drawback, shouldn't the API docs for the configuration call this out?
> 
> Are batches only used when user code manually calls a batch method? Or does the AWS library decide to batch some things? I'm curious if two requests to a Node server come in at the same time, one with request ID 1, one with request ID 2, and both making an outbound AWS request. Does the AWS SDK internally every combine those into a batch? And if so, would the request IDs be properly associated?

The drawbacks are mainly performance related. APM already adds a good amount of overhead, and injecting every message (batches are limited to 10 messages) per API call, especially if they are doing multiple/many batch API calls could push overhead higher than some customers want. There's also size limitations/concerns, and generally messaging services like SQS will charge per set amount of data in a message (64 kb = 1 "SQS request"), so increasing message size will increase customer cost. 

After typing all that out, I think it's best we keep this configurable.

For your second question, batch operations are specifically called by the user. The library does not try to batch requests together. So two separate `SQS.sendMessage` operations sending 1 message each will result in 2 API calls.

**Review by Qard — APPROVED** [review]

**Comment by penchef** [comment]
🙌

**Pull request merged**